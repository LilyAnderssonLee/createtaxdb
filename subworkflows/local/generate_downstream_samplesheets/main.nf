//
// Subworkflow with functionality specific to the nf-core/createtaxdb pipeline
//

workflow GENERATE_DOWNSTREAM_SAMPLESHEETS {
    take:
    ch_databases

    main:
    format     = 'csv' // most common format in nf-core
    format_sep = ','

    // TODO --
    // Make your samplesheet channel construct here depending on your downstream
    // pipelines
    if ( params.generate_pipeline_samplesheets == 'taxprofiler' ) {
        format = 'csv'
        format_sep = ','
        ch_list_for_samplesheet = ch_databases
                                    .map {
                                        meta, db ->
                                            def tool      = meta.tool
                                            def db_name   = meta.id + '-' + meta.tool
                                            def db_params = ""
                                            def db_type   = ""
                                            def db_path   = file(params.outdir).toString() + '/' + meta.tool + '/' + db.getName()
                                        [ tool: tool, db_name: db_name, db_params: db_params, db_type: db_type, db_path: db_path ]
                                    }
                                    .tap{ ch_header }
        if ( params.build_bracken && params.build_kraken2 ) {
            log.warn("Generated nf-core/taxprofiler samplesheet will only have a row for bracken. If Kraken2 is wished to be executed separately, duplicate row and update tool column to Kraken2!")
        }
    }
    // -- FINISH TODO

    // Constructs the header string and then the strings of each row, and
    // finally concatenates for saving.
    ch_header
        .first()
        .map{ it.keySet().join(format_sep) }
        .concat( ch_list_for_samplesheet.map{ it.values().join(format_sep) })
        .collectFile(
            name:"${params.outdir}/downstream_samplesheet/${params.generate_pipeline_samplesheets}.${format}",
            newLine: true,
            sort: false
        )

}
