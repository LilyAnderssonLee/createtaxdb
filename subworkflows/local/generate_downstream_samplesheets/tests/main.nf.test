nextflow_workflow {

    name "Test Subworkflow GENERATE_DOWNSTREAM_SAMPLESHEETS"
    script "../main.nf"
    workflow "GENERATE_DOWNSTREAM_SAMPLESHEETS"

    tag "subworkflows"
    tag "subworkflows_local"
    tag "subworkflows/generate_downstream_samplesheets"

    test("databases - taxprofiler") {

        when {
            params {
                outdir                          = "$outputDir"
                generate_pipeline_samplesheets  = 'taxprofiler'
            }
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [id:'database', tool:'krakenuniq'],
                        file('/workspace/createtaxdb/work/af/59cb1f95d809c3de6520bb0c0d0f6c/database')
                    ],
                    [
                        [id:'database', tool:'centrifuge'],
                        file('/workspace/createtaxdb/work/59/3a1df5c9f4b84a4ec3c8014aa13774/database')
                    ],
                    [
                        [id:'database', tool:'kaiju'],
                        file('/workspace/createtaxdb/work/87/865072d7fce3bbf8c5dfd867d16064/database.fmi')
                    ],
                    [
                        [id:'database', tool:'diamond'],
                        file('/workspace/createtaxdb/work/1b/f20ef16fbc0f4708db40d76410be5c/database.dmnd')
                    ],
                    [
                        [id:'database', tool:'malt'],
                        file('/workspace/createtaxdb/work/4f/c630a9baab2c2f31f15fab1feec4f2/malt_index')
                    ]
                )
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot("${params.outdir}/downstream_samplesheets/${params.generate_pipeline_samplesheets}.csv").match() }
            )
        }
    }
}
