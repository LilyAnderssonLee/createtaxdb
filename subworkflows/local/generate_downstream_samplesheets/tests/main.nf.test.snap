{
    "databases - taxprofiler": {
        "content": [
            "taxprofiler.csv:md5,dc4cb1c64d2d48273eb2125266b315bf"
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-10T22:10:48.802711152"
    }
}